File: legal-contract-pipeline/backend/app/graph/state.py
"""Typed LangGraph state — every agent reads from and writes to this schema."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


# ── Per-clause extracted data ────────────────────────────────────────────────

class ClauseExtract(BaseModel):
    id: int
    type: str
    heading: str
    text: str
    page_hint: float = 0.0  # 0.0–1.0 position in document


class ClauseRiskScore(BaseModel):
    clause_id: int
    score: int = Field(ge=0, le=100)
    severity: Literal["low", "medium", "high", "critical"]
    reasoning: str
    flags: list[str] = Field(default_factory=list)
    rag_hits: list[str] = Field(default_factory=list)  # Pinecone IDs


class ComplianceViolation(BaseModel):
    framework: str
    article: str = ""
    description: str


class ComplianceResult(BaseModel):
    clause_id: int
    compliant: bool
    violations: list[dict[str, Any]] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)


class ChangeRecord(BaseModel):
    original: str
    replacement: str
    rationale: str


class RedlineEdit(BaseModel):
    clause_id: int
    original_text: str
    revised_text: str
    changes: list[dict[str, Any]] = Field(default_factory=list)
    attorney_note: str = ""


# ── Final aggregated report ──────────────────────────────────────────────────

class FinalReport(BaseModel):
    contract_id: str
    contract_name: str
    overall_score: int = Field(ge=0, le=100)
    clause_count: int
    high_risk_count: int
    violation_count: int
    executive_summary: str
    risk_scores: list[ClauseRiskScore]
    compliance_results: list[ComplianceResult]
    redline_edits: list[RedlineEdit]
    generated_at: datetime = Field(default_factory=datetime.utcnow)


# ── Human-in-the-loop approval ───────────────────────────────────────────────

class ApprovalDecision(BaseModel):
    approved: bool
    reviewer_id: str
    reviewer_notes: str = ""
    decided_at: datetime = Field(default_factory=datetime.utcnow)


# ── Full pipeline state ──────────────────────────────────────────────────────

class ContractReviewState(BaseModel):
    """Immutable-ish typed state passed between all agents in the graph."""

    # Input
    contract_id: str
    contract_name: str = ""
    contract_text: str

    # Stage outputs (populated incrementally)
    clauses: list[ClauseExtract] = Field(default_factory=list)
    clause_count: int = 0

    risk_scores: list[ClauseRiskScore] = Field(default_factory=list)

    compliance_results: list[ComplianceResult] = Field(default_factory=list)
    compliance_violation_count: int = 0

    redline_edits: list[RedlineEdit] = Field(default_factory=list)

    report: FinalReport | None = None
    pipeline_complete: bool = False

    # Human-in-the-loop
    awaiting_approval: bool = False
    approval: ApprovalDecision | None = None

    # Metadata
    user_id: str = ""
    created_at: datetime = Field(default_factory=datetime.utcnow)
    error: str | None = None

    class Config:
        # Allow arbitrary types for LangGraph compatibility
        arbitrary_types_allowed = True

File: pipeline-worker/src/pipeline/state.ts
export interface ClauseExtract {
  id: number;
  type: string;
  heading: string;
  text: string;
  page_hint?: number;
}

export interface ClauseRiskScore {
  clause_id: number;
  score: number;
  severity: 'low' | 'medium' | 'high' | 'critical';
  reasoning: string;
  flags: string[];
  rag_hits?: string[];
}

export interface ComplianceViolation {
  framework: string;
  article?: string;
  description: string;
  confidence?: number;
}

export interface ComplianceResult {
  clause_id: number;
  compliant: boolean;
  violations: ComplianceViolation[];
  recommendations: string[];
}

export interface RedlineEdit {
  clause_id: number;
  original_text: string;
  revised_text: string;
  changes: Array<{
    type: string;
    original: string;
    replacement: string;
    rationale: string;
  }>;
  attorney_note?: string;
  approved?: boolean | null;
  original_score?: number;
  mitigated_score?: number;
  mitigation_score_mismatch?: boolean;
}

export interface FinalReport {
  contract_id: string;
  contract_name: string;
  overall_score: number;
  clause_count: number;
  high_risk_count: number;
  violation_count: number;
  compliance_violation_clauses: number;
  compliance_violation_issues: number;
  executive_summary: string;
  risk_scores: ClauseRiskScore[];
  compliance_results: ComplianceResult[];
  redline_edits: RedlineEdit[];
  represented_party: string;
  extraction_integrity_warning?: string | null;
  generated_at?: Date;
}

export interface ContractReviewState {
  contract_id: string;
  contract_name: string;
  contract_text: string;
  represented_party: string;

  clauses: ClauseExtract[];
  clause_count?: number;
  extraction_integrity_warning?: string | null;

  risk_scores: ClauseRiskScore[];

  compliance_results: ComplianceResult[];
  compliance_violation_count?: number;
  compliance_violation_clauses?: number;
  compliance_violation_issues?: number;

  redline_edits: RedlineEdit[];

  report?: FinalReport | null;
  pipeline_complete?: boolean;

  awaiting_approval?: boolean;
  user_id: string;
}
